### ISSUE TITLE
SQLiteDatabaseLockedException: database is locked (code 5)

### ISSUE THREAD

**Issue by KushGoyal** [greenrobot/greenDAO#191]
Some of my users are facing SQLiteDatabaseLockedException.

I am using this code in my application class onCreate method to setup database:

```
private void setupDatabase() {
        DbUpdateHelper helper = new DbUpdateHelper(this, "app-name-db", null);
        SQLiteDatabase db = helper.getWritableDatabase();
        DaoMaster daoMaster = new DaoMaster(db);
        daoSession = daoMaster.newSession();
    }
```

This is the only place where I am accessing the db using helper.getWritableDatabase().

Everywhere else I am using the daoSession object to access db.

Stacktrace:

```
java.lang.RuntimeException: Unable to create application o.ᴛ: android.database.sqlite.SQLiteDatabaseLockedException: database is locked (code 5): , while compiling: PRAGMA journal_mode
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4608)
       at android.app.ActivityThread.access$1500(ActivityThread.java:148)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1353)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:135)
       at android.app.ActivityThread.main(ActivityThread.java:5312)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:901)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:696)
Caused by: android.database.sqlite.SQLiteDatabaseLockedException: database is locked (code 5): , while compiling: PRAGMA journal_mode
       at android.database.sqlite.SQLiteConnection.nativePrepareStatement(SQLiteConnection.java)
       at android.database.sqlite.SQLiteConnection.acquirePreparedStatement(SQLiteConnection.java:889)
       at android.database.sqlite.SQLiteConnection.executeForString(SQLiteConnection.java:634)
       at android.database.sqlite.SQLiteConnection.setJournalMode(SQLiteConnection.java:320)
       at android.database.sqlite.SQLiteConnection.setWalModeFromConfiguration(SQLiteConnection.java:294)
       at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:215)
       at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:193)
       at android.database.sqlite.SQLiteConnectionPool.openConnectionLocked(SQLiteConnectionPool.java:463)
       at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:185)
       at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:177)
       at android.database.sqlite.SQLiteDatabase.openInner(SQLiteDatabase.java:806)
       at android.database.sqlite.SQLiteDatabase.open(SQLiteDatabase.java:791)
       at android.database.sqlite.SQLiteDatabase.openDatabase(SQLiteDatabase.java:694)
       at android.app.ContextImpl.openOrCreateDatabase(ContextImpl.java:1276)
       at android.content.ContextWrapper.openOrCreateDatabase(ContextWrapper.java:267)
       at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:223)
       at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:163)
       at o.ᴛ.onCreate(SourceFile:1032)
       at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1034)
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4605)
       at android.app.ActivityThread.access$1500(ActivityThread.java:148)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1353)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:135)
       at android.app.ActivityThread.main(ActivityThread.java:5312)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:901)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:696)
```

**Comment by Groxx** [comment]
It probably means you're creating multiple DbUpdateHelper instances.  You almost always only want to make one, ever.

**Comment by KushGoyal** [comment]
Hi,

Thanks for the help. Will using a singleton pattern solve this bug?

**Comment by Groxx** [comment]
If that's the cause, yes.  I think you can (and should) just make a single DaoSession as well, so you can just make the session into a singleton.  I forget exactly what goes into DaoSession though, so I could be wrong on that - check the docs :)

**Comment by KushGoyal** [comment]
I think I cannot make DbUpdateHelper singleton because the static instance will hold a reference to the context which will not get garbage collect until classloader has been unloaded which is I think almost never.

Is there any other solution I can implement to prevent multiple instances of DbUpdateHelper. Although I still don't understand how is it even possible that there are multiple instances of DbUpdateHelper since I am using the code in the application class and not anywhere else.

**Comment by zhangsl** [comment]
I got a similar crash, I have a singleton DatabaseHelper, and no transaction operation, but still crash

**Comment by hwldzh** [comment]
you  can use context.getapplicationcontext, then the context can be collect by the garbage

**Comment by michal-luszczuk** [comment]
Did you solve this issue?
Because i'm also facing this problem.
App is crashing while first `getWritableDatabase`call. And this call is done inside Application object. So nothing should be called before.

**Comment by donghuipei** [comment]
@michal-luszczuk 
My issue is like you, I found the cause of the problem is a process that would correspond to a target application, if your application is configured with additional process, you might have multiple processes simultaneously open db operation, causing database locks abnormal.

**Comment by michal-luszczuk** [comment]
@donghuipei Yeah, I've got same problem - multi-proces app related (database was initiated in `Application` class `onCreate()` method which was called by each process separately)

**Comment by donghuipei** [comment]
@michal-luszczuk 
If only the main process in the application database operations there, while other applications related to the process does not require operation of the database, I recommend to initialize the database plus operating conditions, like this:
if(getCurProcessName(getApplicationContext()).equals("replace your application package name")){
            // initialize the database
 }

    private String getCurProcessName(Context context) {
        int pid = android.os.Process.myPid();
        ActivityManager activityManager = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        for (ActivityManager.RunningAppProcessInfo appProcess : activityManager.getRunningAppProcesses()) {
            if (appProcess.pid == pid) {
                return appProcess.processName;
            }
        }
        return "";
    }

**Comment by michal-luszczuk** [comment]
@donghuipei 
That's right (I did the same), and probably this is the best solution.
I think this answer will help many devs :)

**Comment by artesa** [comment]
I wonder what's the solution in case if _multiple_ processes need accessing the same SQLite database...

**Comment by a121103086** [comment]
曰了狗了 什么鬼东西

**Comment by greenrobot-team** [comment]
@artesa Based on a quick Google search (\*hint\*) using a ContentProvider is the common solution to access a (SQLite) database from another process/app. -ut

**Comment by artesa** [comment]
@greenrobot-team Clever but nonconstructive, and I can explain why. SQLite documentation states that SQLite should support multiprocess access out-of-box. Using a ContentProvider is basically accessing SQLite database from a *single* process, while accessing the provider itself happens in a multiprocess manner. Therefore, it seems like the problem lies in Google's implementation of SQLiteOpenHelper, which makes multiprocess access done through it unreliable while you keep the connection open (and, of course, there is a plenty of other problems in case you don't).

I reached some progress in struggling with the issue, but still was not able to get rid of it completely. That's why I'm interested whether anyone was more successful.

BTW, I don't believe ContentProviders are actually a panacea. The library based on them which we used was generating even more errors on heavy loads than accessing SQLite directly. Was it the problem of the library? Maybe. But it's rather popular and we failed to find the reason in code, which itself makes providers either unreliable or hard to set up properly in some cases.

As a generalization of your comment I could add the following: based on a quick Google search (*hint*) dividing your app into multiple processes is strictly discouraged in principle. Yep, that's true. But, would this knowledge help me to deal with that huge ton of legacy? Don't think so. :)

**Comment by hooby3dfx** [comment]
I am also seeing this happen occasionally in an app that I work on, but the app is not designed to use multiple processes. 
That being the case, when would this happen? Would that PID check still be useful?

**Comment by nickylin** [comment]
I'm occurred this problem, so we do not use greendao for multi process?
`

Thread Name: 'ModelWriteWorkThread'
--
  |   | Back traces starts.
  |   | android.database.sqlite.SQLiteDatabaseLockedException: database is locked (code 5)
  |   | at android.database.sqlite.SQLiteConnection.nativeExecuteForChangedRowCount(Native Method)
  |   | at android.database.sqlite.SQLiteConnection.executeForChangedRowCount(SQLiteConnection.java:915)
  |   | at android.database.sqlite.SQLiteSession.executeForChangedRowCount(SQLiteSession.java:755)
  |   | at android.database.sqlite.SQLiteStatement.executeUpdateDelete(SQLiteStatement.java:64)
  |   | at android.database.sqlite.SQLiteDatabase.executeSql(SQLiteDatabase.java:1771)

`

**Comment by xurui1995** [comment]
@michal-luszczuk 
I got the exception and i didnt use GreenDao. 
Multiple processes were used in my app, but only one processes used Sqlite. 
have u solve it?

**Comment by michal-luszczuk** [comment]
@xurui1995 
Like I said before. Problem is **solved for us.**
The case was related to opening/initializing greenDao/sqlite db in code which was called by multiple processes.

You should check in which process code is called (using [This comment](https://github.com/greenrobot/greenDAO/issues/191#issuecomment-235522351) ) and this should resolve your problems.

**Comment by nickylin** [comment]
if we need to read or write db in multi process, how we do to solve this problem? @michal-luszczuk

**Comment by zxy198717** [comment]
if we need to read or write db in multi process, how we do to solve this problem? @michal-luszczuk @donghuipei

**Comment by EnvySongtao** [comment]
I have the similar crash，just becouse,I write self Application but forget to set appliaction name in manifest.xml

**Comment by abduldblogger** [comment]
Hi,
Use : 
database = dbHelper.getWritableDatabase();
            database.enableWriteAheadLogging(); // this will allow transactions from multiple threads.

Please read https://developer.android.com/reference/android/database/sqlite/SQLiteDatabase.html#enableWriteAheadLogging%28%29

I hope this will help.

**Comment by ashaevy** [comment]
Found how to solve it. The problem for me was that I created multiple instances of SQLiteOpenHelper for the same database file. getWritableDatabase() contains synchronized block inside to prevent parallel access from multiple threads but because as I said there were multiple instances of SQLiteOpenHelper this lock didn't work. I created singleton from SQLiteOpenHelper. In order to release memory, I used WeakReference for "instance" field.
